Mathlib/Algebra/Category/ModuleCat/Ulift.lean

```diff
@@ -0,0 +1,96 @@
+/-
+Copyright (c) 2025 Nailin Guan. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Nailin Guan
+-/
+module
+
+public import Mathlib.Algebra.Category.ModuleCat.Injective
+public import Mathlib.Algebra.Category.ModuleCat.Projective
+public import Mathlib.Algebra.Homology.ShortComplex.ModuleCat
+public import Mathlib.CategoryTheory.Linear.LinearFunctor
+public import Mathlib.CategoryTheory.Preadditive.Injective.Preserves
+public import Mathlib.CategoryTheory.Preadditive.Projective.Preserves
+
+/-!
+
+# Ulift Functor of ModuleCat
+
+In this file, we define the obvious functor `ModuleCat.{v} R ⥤ ModuleCat.{max v v'} R` and prove
+it is exact, fully faithful and preverves projective and injective objects.
+
+-/
+
+@[expose] public section
+
+universe v' v u
+
+variable (R : Type u)
+
+open CategoryTheory
+
+namespace ModuleCat
+
+section Ring
+
+variable [Ring R]
+
+/-- Universe lift functor for `R`-module. -/
+@[simps obj map]
+def uliftFunctor : ModuleCat.{v} R ⥤ ModuleCat.{max v v'} R where
+  obj X := ModuleCat.of R (ULift.{v', v} X)
+  map f := ModuleCat.ofHom <|
+    ULift.moduleEquiv.symm.toLinearMap.comp (f.hom.comp ULift.moduleEquiv.toLinearMap)
+
+/-- The universe lift functor for `R`-module is fully faithful. -/
+def fullyFaithfulUliftFunctor : (uliftFunctor R).FullyFaithful where
+  preimage f := ModuleCat.ofHom (ULift.moduleEquiv.toLinearMap.comp
+    (f.hom.comp ULift.moduleEquiv.symm.toLinearMap))
+
+instance : (uliftFunctor.{v', v} R).Full := (fullyFaithfulUliftFunctor R).full
+
+instance : (uliftFunctor.{v', v} R).Faithful := (fullyFaithfulUliftFunctor R).faithful
+
+instance : (uliftFunctor R).Additive where
+
+lemma uliftFunctor_map_exact (S : ShortComplex (ModuleCat.{v} R)) (h : S.Exact) :
+    (S.map (uliftFunctor R)).Exact := by
+  rw [CategoryTheory.ShortComplex.ShortExact.moduleCat_exact_iff_function_exact]
+  simp only [uliftFunctor, ShortComplex.map_X₁, ShortComplex.map_X₂, ShortComplex.map_X₃,
+    ShortComplex.map_f, hom_ofHom, LinearMap.coe_comp, LinearEquiv.coe_coe, ShortComplex.map_g]
+  intro x
+  simp only [Function.comp_apply, Set.mem_range, LinearEquiv.symm_apply_eq, map_zero]
+  rw [(CategoryTheory.ShortComplex.ShortExact.moduleCat_exact_iff_function_exact S).mp h]
+  cat_disch
+
+instance : Limits.PreservesFiniteLimits (uliftFunctor.{v', v} R) := by
+  have := ((CategoryTheory.Functor.exact_tfae (uliftFunctor.{v', v} R)).out 1 3).mp
+    (uliftFunctor_map_exact R)
+  exact this.1
```

Comment:
How hard is it to show that `uliftFunctor.{v', v} R` preserves all limits (at least when `Small.{v} R`)? I think the argument is just that the limit of `Small.{v}` modules in `ModuleCat.{max v v'} R` is still `Small.{v}` and by `CreatingLimit` nonsense you are done. I believe the same works for filtered colimits as well. But not for general colimits.

Reply:
I am with @erdOne here: the argument is not very long, and so I think it should be included in this PR:
```lean
instance : Limits.PreservesLimitsOfSize.{v, v} (uliftFunctor.{v', v} R) :=
  haveI :  Limits.PreservesLimitsOfSize.{v, v} (uliftFunctor.{v', v} R ⋙ forget _) := by
    change Limits.PreservesLimitsOfSize.{v, v} (forget _ ⋙ CategoryTheory.uliftFunctor.{v'})
    infer_instance
  Limits.preservesLimits_of_reflects_of_preserves (uliftFunctor.{v', v} R) (forget _)

instance : Limits.PreservesFiniteLimits (uliftFunctor.{v', v} R) :=
  Limits.PreservesLimitsOfSize.preservesFiniteLimits _
```